service/storage_proxy.cc

```diff
@@ -1817,11 +1817,32 @@ storage_proxy::mutate_counter_on_leader_and_replicate(const schema_ptr& s, froze
     auto shard = _db.local().shard_of(fm);
     bool local = shard == this_shard_id();
     get_stats().replica_cross_shard_ops += !local;
-    return _db.invoke_on(shard, {_write_smp_service_group, timeout}, [gs = global_schema_ptr(s), fm = std::move(fm), cl, timeout, gt = tracing::global_trace_state_ptr(std::move(trace_state)), permit = std::move(permit), local] (database& db) {
+    bool preimage_needed = s->cdc_options().preimage();
+    return _db.invoke_on(shard, {_write_smp_service_group, timeout}, [this, gs = global_schema_ptr(s), fm = std::move(fm), cl, timeout, gt = tracing::global_trace_state_ptr(std::move(trace_state)), permit = std::move(permit), local, preimage_needed] (database& db) {
         auto trace_state = gt.get();
         auto p = local ? std::move(permit) : /* FIXME: either obtain a real permit on this shard or hold original one across shard */ empty_service_permit();
-        return db.apply_counter_update(gs, fm, timeout, trace_state).then([cl, timeout, trace_state, p = std::move(p)] (mutation m) mutable {
-            return service::get_local_storage_proxy().replicate_counter_from_leader(std::move(m), cl, std::move(trace_state), timeout, std::move(p));
+        return db.apply_counter_update(gs, fm, timeout, trace_state, preimage_needed).then([cl, timeout, trace_state, p = std::move(p), delta = fm.unfreeze(gs)]
+                (std::tuple<mutation, mutation_opt>&& state_with_opt_preimage) mutable {
+            auto new_state = std::get<0>(std::move(state_with_opt_preimage));
+            auto old_state = std::get<1>(std::move(state_with_opt_preimage));
+            
+            future<> cdc_fut = make_ready_future<>();
+            auto cdc_ptr = service::get_local_storage_proxy()._cdc;
+            if (cdc_ptr && cdc_ptr->counter_needs_cdc_augmentation({delta})) {
+                cdc_fut = cdc_ptr->calc_counter_mutation_augmentation(timeout, std::move(delta), std::move(old_state), trace_state).then(
+                        [cl, timeout, trace_state, p](std::tuple<std::vector<mutation>, lw_shared_ptr<cdc::operation_result_tracker>> t) mutable {
+                    auto cdc_mutations = std::move(std::get<0>(t));
+                    if (cdc_mutations.empty()) {
+                        return make_ready_future<>();
+                    }
+                    auto cdc_tracker = std::move(std::get<1>(t));
+                    return service::get_local_storage_proxy().mutate_internal(std::move(cdc_mutations), cl, false, trace_state, p, timeout, std::move(cdc_tracker));
```

Comment:
So we perform 2 round trips: one to apply the cdc mutation, the other to apply the counter mutation.
Can't we attach the cdc mutation to the `replicate_counter_from_leader` path?

Reply:
There was such an attempt (and this patch consisted of ~10 LOC), but augmenting counters needs the context of counter tombstones. This context is reliably available only at the moment of counter update. Otherwise we cannot get rid of updates of dead counters in cdc log.

Reply:
I don't see how those two things are related.
There is a moment where you have two (sets of) mutations ready to be sent: `new_state`, and `cdc_mutations`. I'm simply suggesting to send these two mutations in a single vector / in parallel, so we don't do 2 round trips.